sql/sql_path.h

```diff
@@ -0,0 +1,145 @@
+#ifndef SQL_PATH_INCLUDED
+#define SQL_PATH_INCLUDED
+
+#include "mysqld.h"
+#include "sql_array.h"
+#include "sql_list.h"
+
+
+/* forward declarations */
+class Database_qualified_name;
+
+
+struct Sql_path
+{
+private:
+  LEX_CSTRING m_schemas_array[16];
+  Bounds_checked_array<LEX_CSTRING> m_schemas;
+  uint m_count;
+  char *m_buffer;
+  size_t m_buffer_length;
+
+public:
+  Sql_path();
+  Sql_path& operator=(const Sql_path &rhs)
+  {
+    set(rhs, 1);
+    return *this;
+  }
+  Sql_path& operator=(Sql_path &&rhs)
+  {
+    set(std::move(rhs), 1);
+    return *this;
+  }
+  Sql_path(const Sql_path&) = delete;
+
+  void init_array()
+  {
+    m_schemas.reset(m_schemas_array,
+                    sizeof(m_schemas_array) / sizeof(m_schemas_array[0]));
+  }
+
+  bool find_db_unqualified(THD *thd, const LEX_CSTRING &name,
+                           const Sp_handler *sph,
+                           Lex_ident_db_normalized *dbn_out,
+                           sp_name **spname_out) const;
+  bool find_db_qualified(THD *thd,
+                         sp_name *name,
+                         const Sp_handler **sph,
+                         Database_qualified_name *pkgname) const;
```

Comment:
I know by now what you mean by "qualified" and "unqualified", but it's still confusing. May be better rename (and reword everywhere in the commit) to something like find_db_routine and find_db_package_routine ?